File: src/main/java/net/Chidoziealways/everythingjapanese/datagen/DataGenerators.java
package net.Chidoziealways.everythingjapanese.datagen;

import com.google.common.collect.ImmutableMap;
import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.minecraft.Util;
import net.minecraft.core.*;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.data.DataGenerator;
import net.minecraft.data.DataProvider;
import net.minecraft.data.PackOutput;
import net.minecraft.data.loot.LootTableProvider;
import net.minecraft.data.registries.VanillaRegistries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.biome.Biome;
import net.minecraft.world.level.levelgen.placement.BiomeFilter;
import net.minecraft.world.level.levelgen.placement.PlacedFeature;
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
import net.minecraftforge.common.data.BlockTagsProvider;
import net.minecraftforge.common.data.ExistingFileHelper;
import net.minecraftforge.common.data.ForgeAdvancementProvider;
import net.minecraftforge.data.event.GatherDataEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiFunction;

@Mod.EventBusSubscriber(modid = EverythingJapanese.MOD_ID, bus = Mod.EventBusSubscriber.Bus.MOD)
public class DataGenerators {

    @SubscribeEvent
    public static void gatherData(GatherDataEvent event){
        DataGenerator generator = event.getGenerator();
        PackOutput packOutput = generator.getPackOutput();
        ExistingFileHelper existingFileHelper = event.getExistingFileHelper();
        CompletableFuture<HolderLookup.Provider> lookupProvider = event.getLookupProvider();
        CompletableFuture<HolderLookup.Provider> fLookupProvider = CompletableFuture.supplyAsync(ModDatapackEntries::createLookup, Util.backgroundExecutor());

        generator.addProvider(event.includeServer(), new LootTableProvider(packOutput, Collections.emptySet(),
                List.of(new LootTableProvider.SubProviderEntry(ModBlockLootTableProvider::new, LootContextParamSets.BLOCK)), lookupProvider));

        BlockTagsProvider blockTagsProvider = new ModBlockTagProvider(packOutput, lookupProvider, existingFileHelper);
        generator.addProvider(event.includeServer(), blockTagsProvider);
        generator.addProvider(event.includeServer(), new ModItemTagProvider(packOutput, lookupProvider, blockTagsProvider.contentsGetter(), existingFileHelper));
        generator.addProvider(event.includeServer(), new ModBiomeTagProvider(packOutput, lookupProvider,  existingFileHelper));

        generator.addProvider(event.includeServer(), new ModDatapackEntries(packOutput, lookupProvider));

        generator.addProvider(event.includeClient(), new ModModelProvider(packOutput));

        generator.addProvider(event.includeServer(), new ModGlobalLootModifierProvider(packOutput, lookupProvider));

        generator.addProvider(event.includeServer(), new ModRecipeProvider.Runner(packOutput, fLookupProvider));

        generator.addProvider(event.includeServer(), ModAdvancementProvider.create(packOutput, fLookupProvider, existingFileHelper));

    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/datagen/DataGenerators.kt
package net.Chidoziealways.everythingjapanese.datagen

import net.Chidoziealways.everythingjapanese.JAPANESE_MOD_ID
//import net.Chidoziealways.everythingkorean.datagen.KModDatapackEntries
import net.minecraft.Util
import net.minecraft.core.HolderLookup
import net.minecraft.data.loot.LootTableProvider
import net.minecraft.data.loot.LootTableProvider.SubProviderEntry
import net.minecraft.data.registries.RegistryPatchGenerator
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.neoforge.data.event.GatherDataEvent
import thedarkcolour.kotlinforforge.common.KotlinMod
import java.util.concurrent.CompletableFuture

@KotlinMod.KotlinEventBusSubscriber(modId = JAPANESE_MOD_ID)
object DataGenerators {
    @SubscribeEvent
    fun gatherClientData(event: GatherDataEvent.Client) {
        // Pass this factory to your RegistrySetBuilder or generator
        val generator = event.generator
        val packOutput = generator.packOutput

        val lookupFuture = RegistryPatchGenerator.createLookup(
            event.lookupProvider,
            JModDatapackEntries.BUILDER
        )/*.thenCompose { patched1 ->
            RegistryPatchGenerator.createLookup(
                CompletableFuture.completedFuture(patched1.full()),
                KModDatapackEntries.BUILDER
            )
        }*/

        val lookupProvider =
            lookupFuture.thenApply { it.full() }

        generator.addProvider(
            true,
            EnglishLanguageProvider(
                packOutput
            )
        )

        event.createProvider(::ModDataMapProvider)

        generator.addProvider(
            true,
            JapaneseLanguageProvider(
                packOutput
            )
        )

        generator.addProvider(
            true,
            KoreanLanguageProvider(
                packOutput
            )
        )

        generator.addProvider(
            true,
            SimplifiedChineseLanguageProvider(
                packOutput
            )
        )

        generator.addProvider(
            true,
            TraditionalChineseLanguageProvider(
                packOutput
            )
        )

        generator.addProvider(
            true,
            LootTableProvider(
                packOutput,
                mutableSetOf(),
                listOf(
                    SubProviderEntry({ registries ->
                        ModBlockLootTableProvider(registries)
                    }, LootContextParamSets.BLOCK),
                    SubProviderEntry( {registries ->
                        ModChestLootTableProvider(registries)
                    }, LootContextParamSets.CHEST)
                ),
                lookupProvider
            )
        )

        generator.addProvider(
            true
        ) { JModDatapackEntries(packOutput, lookupProvider) }
        generator.addProvider(
            true,
            ModBlockTagProvider(packOutput, lookupProvider)
        )
        generator.addProvider(
            true,
            ModItemTagProvider(packOutput, lookupProvider)
        )
        event.createProvider(::ModEntityTagProvider)
        generator.addProvider(
            true,
            ModBiomeTagProvider(packOutput, lookupProvider)
        )
        generator.addProvider(
            true,
            ModFluidTagProvider(packOutput, lookupProvider)
        )

        generator.addProvider(true, ModModelProvider(packOutput, lookupProvider))

        generator.addProvider(
            true,
            ModGlobalLootModifierProvider(packOutput, lookupProvider)
        )

        generator.addProvider(
            true,
            ModRecipeProvider.Runner(packOutput, lookupProvider)
        )

        generator.addProvider(
            true,
            ModAdvancementProvider.create(packOutput, lookupProvider)
        )
    }

    @SubscribeEvent
    fun gatherServerData(event: GatherDataEvent.Server) {
        // Pass this factory to your RegistrySetBuilder or generator
        val generator = event.generator
        val packOutput = generator.packOutput

        val lookupProvider = event.lookupProvider
        val fLookupProvider = CompletableFuture.supplyAsync(
            { JModDatapackEntries.createLookup() },
            Util.backgroundExecutor()
        )

        generator.addProvider(
            true,
            LootTableProvider(
                packOutput,
                mutableSetOf(),
                listOf(
                    SubProviderEntry({ registries ->
                        ModBlockLootTableProvider(registries)
                    }, LootContextParamSets.BLOCK),
                    SubProviderEntry( {registries ->
                        ModChestLootTableProvider(registries)
                    }, LootContextParamSets.CHEST)
                ),
                lookupProvider
            )
        )

        generator.addProvider(
            true,
            ModBlockTagProvider(packOutput, lookupProvider)
        )
        generator.addProvider(
            true,
            ModItemTagProvider(packOutput, lookupProvider)
        )
        generator.addProvider(
            true,
            ModBiomeTagProvider(packOutput, fLookupProvider)
        )

        generator.addProvider(
            true,
            JModDatapackEntries(packOutput, fLookupProvider)
        )

        generator.addProvider(true, ModModelProvider(packOutput, fLookupProvider))

        generator.addProvider(
            true,
            ModGlobalLootModifierProvider(packOutput, lookupProvider)
        )

        generator.addProvider(
            true,
            ModRecipeProvider.Runner(packOutput, fLookupProvider)
        )

        generator.addProvider(
            true,
            ModAdvancementProvider.create(packOutput, fLookupProvider)
        )
    }
}